crates/findex/src/implementations/redis/findex.rs

```diff
@@ -0,0 +1,434 @@
+            num_reindexing_before_full_set,
+        )
+        .await?;
+        Ok(())
+    }
+}
+
+impl FindexCallbacks<FindexRedisError, UID_LENGTH> for FindexRedis {
+    async fn progress(
+        &self,
+        _results: &HashMap<Keyword, HashSet<IndexedValue>>,
+    ) -> Result<bool, FindexRedisError> {
+        //TODO: allow passing callback fn on connect
+        Ok(true)
+    }
+
+    async fn fetch_all_entry_table_uids(&self) -> Result<Uids<UID_LENGTH>, FindexRedisError> {
+        let keys: Vec<Vec<u8>> = self.mgr.clone().keys(key(FindexTable::Entry, b"*")).await?;
+        trace!("fetch_all_entry_table_uids num keywords: {}", keys.len());
+        Ok(Uids(
+            keys.iter()
+                .map(|v| {
+                    let mut uid = [0u8; UID_LENGTH];
+                    uid.copy_from_slice(&v[4..]);
+                    Uid::from(uid)
+                })
+                .collect(),
+        ))
+    }
+
+    async fn fetch_entry_table(
+        &self,
+        entry_table_uids: Uids<UID_LENGTH>,
+    ) -> Result<EncryptedMultiTable<UID_LENGTH>, FindexRedisError> {
+        trace!(
+            "fetch_entry_table num keywords: {}:",
+            entry_table_uids.0.len(),
+        );
+        // guard against empty uids
+        if entry_table_uids.0.is_empty() {
+            return Ok(EncryptedMultiTable::default());
+        }
+
+        // build Redis keys
+        let keys: Vec<Vec<u8>> = entry_table_uids
+            .0
+            .iter()
+            .map(|uid| key(FindexTable::Entry, uid))
+            .collect();
+
+        // mget the values from the Redis keys
+        let values: Vec<Vec<u8>> = self.mgr.clone().mget(keys).await?;
+
+        // discard empty values
+        let tuples = entry_table_uids
+            .0
+            .into_iter()
+            .zip(values)
+            .filter(|(_uid, v)| !v.is_empty())
+            .collect::<Vec<_>>();
+        trace!("fetch_entry_table non empty tuples len: {}", tuples.len(),);
+
+        Ok(EncryptedMultiTable(tuples))
+    }
+
+    async fn fetch_chain_table(
+        &self,
+        chain_table_uids: Uids<UID_LENGTH>,
+    ) -> Result<EncryptedTable<UID_LENGTH>, FindexRedisError> {
+        trace!(
+            "fetch_chain_table num entries: {}:",
+            chain_table_uids.0.len(),
+        );
+
+        //guard against empty uids
+        if chain_table_uids.0.is_empty() {
+            return Ok(EncryptedTable::default());
+        }
+
+        let keys: Vec<Vec<u8>> = chain_table_uids
+            .0
+            .iter()
+            .map(|uid| key(FindexTable::Chain, uid))
+            .collect();
+        let values: Vec<Vec<u8>> = self.mgr.clone().mget(keys).await?;
+
+        Ok(EncryptedTable::from(
+            chain_table_uids
+                .0
+                .into_iter()
+                .zip(values)
+                .filter(|(_uid, v)| !v.is_empty())
+                .collect::<HashMap<Uid<UID_LENGTH>, Vec<u8>>>(),
+        ))
+    }
```

Comment:
Duplicate code, we could have a `fetch` function inside the `FindexRedis` implementation (outside the `FindexCallbacks` impl) which takes UIDs and `FindexTable` and call it.

Reply:
Since these functions do not return the same type, the amount of deduplication will be very small at the expense of readability. 
Keeping as such

Reply:
Yes good point! Didn't remember the change about return types :-(